lib/ecto/changeset.ex

```diff
@@ -2076,23 +2083,34 @@ defmodule Ecto.Changeset do
       ...>   end)
       ...> end)
       %{title: ["should be at least 3 characters"]}
+
+  Optionally, you can pass function that accepts second argument `field` that will contain all
+  validation rules applied to it.
   """
-  @spec traverse_errors(t, (error -> String.t)) :: %{atom => [String.t]}
-  def traverse_errors(%Changeset{errors: errors, changes: changes, types: types}, msg_func)
-      when is_function(msg_func, 1) do
+  @spec traverse_errors(t, (error -> String.t) | (error, Keyword.t -> String.t)) :: %{atom => [String.t]}
+  def traverse_errors(%Changeset{errors: errors, changes: changes, types: types, validations: validations}, msg_func)
+      when is_function(msg_func, 1) or is_function(msg_func, 2) do
     errors
     |> Enum.reverse()
-    |> merge_error_keys(msg_func)
+    |> merge_error_keys(msg_func, validations)
     |> merge_related_keys(changes, types, msg_func)
   end
 
-  defp merge_error_keys(errors, msg_func) do
+  defp merge_error_keys(errors, msg_func, _) when is_function(msg_func, 1)  do
     Enum.reduce(errors, %{}, fn({key, val}, acc) ->
       val = msg_func.(val)
       Map.update(acc, key, [val], &[val|&1])
     end)
   end
 
+  defp merge_error_keys(errors, msg_func, validations) when is_function(msg_func, 2)  do
+    Enum.reduce(errors, %{}, fn({key, val}, acc) ->
+      field = Keyword.get_values(validations, key)
+      val = msg_func.(val, field)
```

Comment:
@AndrewDryga oh, that's true. Good call.

We should pass then `msg_func.(changeset, key, value)`. This way you should have everything you need to retrieve it. I don't want to pass validations because if people need more information in the future, like the changes field, we will end-up with `changes, validations, value` and whatever else they may need.